src/snowflake/snowpark/types.py

```diff
@@ -552,29 +564,41 @@ def __init__(
         column_identifier: Union[ColumnIdentifier, str],
         datatype: DataType,
         nullable: bool = True,
+        is_column: bool = True,
     ) -> None:
-        self.column_identifier = (
-            ColumnIdentifier(column_identifier)
-            if isinstance(column_identifier, str)
-            else column_identifier
-        )
+        self.name = column_identifier
+        self.is_column = is_column
         self.datatype = datatype
         self.nullable = nullable
 
     @property
     def name(self) -> str:
-        """Returns the column name."""
-        return self.column_identifier.name
+        return self.column_identifier.name if self.is_column else self._name
 
```

Comment:
does `is_column=True` enforce type of `self.column_identifier` is `ColumnIdentifier`? can we add an assert for that in `__init__`

Reply:
`__init__` uses the setter for `self.name` which ensures that `self.column_identifier` is always `ColumnIdentifier`.